optaplanner-core/src/test/java/org/optaplanner/core/api/score/stream/AdvancedGroupByConstraintStreamTest.java

```diff
@@ -71,80 +86,155 @@ public void collectedAndFiltered() {
+                assertMatchWithScore(-1,
+                        asMap(solution.getFirstEntityGroup(), 3, solution.getEntityGroupList().get(1), 3)));
 
         // Incremental
-        Stream.of(entity1, entity2).forEach(entity -> {
-            scoreDirector.beforeEntityRemoved(entity);
-            solution.getEntityList().remove(entity);
-            scoreDirector.afterEntityRemoved(entity);
+        TestdataLavishEntity entity = solution.getFirstEntity();
+        scoreDirector.beforeEntityRemoved(entity);
+        solution.getEntityList().remove(entity);
+        scoreDirector.afterEntityRemoved(entity);
+        assertScore(scoreDirector,
+                assertMatchWithScore(-1,
+                        asMap(solution.getEntityGroupList().get(1), 3)));
+    }
+
+    @Test
+    public void quadGroupByRecollected() {
+        assumeDrools();
+        TestdataLavishSolution solution = TestdataLavishSolution.generateSolution(2, 3, 2, 8);
+
+        InnerScoreDirector<TestdataLavishSolution> scoreDirector = buildScoreDirector((factory) -> {
+            return factory.fromUniquePair(TestdataLavishEntity.class, equal(TestdataLavishEntity::getEntityGroup))
+                    .join(TestdataLavishEntity.class,
+                            equal((a, b) -> a.getEntityGroup(), TestdataLavishEntity::getEntityGroup),
+                            filtering((a, b, c) -> !Objects.equals(a, c) && !Objects.equals(b, c)))
+                    .join(TestdataLavishEntity.class,
+                            equal((a, b, c) -> a.getEntityGroup(), TestdataLavishEntity::getEntityGroup),
+                            filtering((a, b, c, d) -> !Objects.equals(a, d) && !Objects.equals(b, d) && !Objects.equals(c, d)))
+                    // Stream of all unique entity quad tuples that share a group
+                    .groupBy((entityA, entityB, entityC, entityD) -> entityA.getEntityGroup(), countQuad())
+                    .groupBy(toMap((g, c) -> g, (g, c) -> c, Integer::sum))
+                    .penalize(TEST_CONSTRAINT_NAME, SimpleScore.ONE);
         });
-        assertScore(scoreDirector, assertMatchWithScore(-1, solution.getFirstEntityGroup(), 1));
+
+        // From scratch
+        scoreDirector.setWorkingSolution(solution);
+        assertScore(scoreDirector,
+                assertMatchWithScore(-1,
+                        asMap(solution.getFirstEntityGroup(), 12, solution.getEntityGroupList().get(1), 12)));
+
+        // Incremental
+        TestdataLavishEntity entity = solution.getFirstEntity();
+        scoreDirector.beforeEntityRemoved(entity);
+        solution.getEntityList().remove(entity);
+        scoreDirector.afterEntityRemoved(entity);
+        assertScore(scoreDirector,
+                assertMatchWithScore(-1,
+                        asMap(solution.getEntityGroupList().get(1), 12)));
     }
 
     @Test
-    @Ignore("Regrouping not yet supported.")
+    @Ignore
```

Comment:
Still ignored? If so, why?

Reply:
The PR is not yet complete. This will be fixed soon.